Implement ResizeObserver infra, support content box calculations.

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->

Implement the basic infra for the [ResizeObserver](https://drafts.csswg.org/resize-observer/) Web API, and support size calculations for content boxes. 

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #31006 (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

gterzian committed:
implement basic infra for ResizeObserver

gterzian commented:
[Those two](https://github.com/servo/servo/pull/31108#issuecomment-1911589318) appear to be intermittent, how do I mark them as such again?

I have filed an issue for tests that are still failing: https://github.com/servo/servo/issues/31182

The base case appears to work, there is some intermittency which appears related to the delivery of rafs.

gterzian commented:
@mrobinson @jdm @emilio  This one is ready for review. See https://github.com/servo/servo/issues/31182 for the currently failing tests(usually it's one sub-test failing in the test)

Are the [tests failure](https://github.com/servo/servo/actions/runs/7695647188/job/20968986306#step:5:16) related to the switch of the main branch? 

> python: can't open file '/home/runner/work/servo/servo/./python/servo/try_parser.py': [Errno 2] No such file or directory

Loirooriol commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,260 @@
+    match observed_box {
+        ResizeObserverBoxOptions::Content_box |
+        ResizeObserverBoxOptions::Device_pixel_content_box => {
+            target.upcast::<Node>().bounding_content_box_or_zero()
```
Device_pixel_content_box should snap to device pixels.
Also, `bounding_content_box_or_zero` is the wrong choice if there are multiple fragments, we don't want the bounding rect. As per the current spec it should just be the 1st fragment, with the idea of extending to an array with all fragments in the future.

gterzian replied:
Thanks! I've added your comment to https://github.com/servo/servo/issues/31182 because I think some of these might best be addressed as follow-ups, so that this PR can focus more on the overal mechanism and get the event-loop integration right. 

Would you happen to have links to the code where one could get the right layout queries or transform the resulting data into the right format?

mrobinson replied:
It should be possible to use the new query API that's been exposed to implement this correct versus in a followup

gterzian replied:
> use the new query API

Do you mean `content_boxes`? If so, I have added a commit that uses that one.

gterzian commented:
I'm looking a bit more into this raf issue that makes a sub test fail, I guess it would be nice to have the entire `event-loop.html` test pass. Currently failing is one where it expects to have received a raf but it hasn't....

Feel free to review in the meantime...

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,260 @@
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-observedbox>
+    observed_box: ResizeObserverBoxOptions,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-lastreportedsizes>
+    last_reported_sizes: DomRefCell<VecDeque<ResizeObserverSizeImpl>>,
```
Yeah nothing... I was thinking of making this just an option, because only the last one is ever read. What do you think?

emilio replied:
An option is fine as long as we only report one size at a time. you really want to set this to the array we last reported, rather than keeping appending.

gterzian replied:
Ah ok, now I understand: it's the array for multiple fragments for a single reported size, where currently one fragment is used. I thought it was more of a stack of reported sizes... thanks!

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,260 @@
+            observation
+                .last_reported_sizes
+                .borrow_mut()
+                .push_back(size_impl);
```
I'm using the equivalent that isn't a dom object, so that `ResizeObservation` doesn't have to be a dom object either(which would with our codegen require exposing it to JS).

gterzian commented:
Regarding the `event-loop.html` test: it does not receive a raf when it expects to. The tests [acknowledges](https://github.com/servo/servo/blob/9b6c473695e14c1a37dd70325657519b901e7efc/tests/wpt/tests/resize-observer/eventloop.html#L91) this potential issue, and tries to work around it, but I guess in our case this is not enough. I tried moving around the raf code in the script-thread and the test would pass, but I don't think we should look too much into this now. [test0](https://github.com/servo/servo/blob/9b6c473695e14c1a37dd70325657519b901e7efc/tests/wpt/tests/resize-observer/eventloop.html#L45) and [test2](https://github.com/servo/servo/blob/9b6c473695e14c1a37dd70325657519b901e7efc/tests/wpt/tests/resize-observer/eventloop.html#L196) pass, with the exeption of the raf count subtest. test1 seems to test shadow dom things, and it times out.

gterzian commented:
@Loirooriol @emilio Thanks to the both of your for your initial comments, which have been addressed. 

Could I get a comprehensive review please(also perhaps from @mrobinson @jdm if you have time)? 

I think the current mechanism works, with two seperate areas that needs to be addressed:
1. Various size calculations are off, as pointed out by @Loirooriol 
2. We fail tests relying on proper integration with the raf callbacks. 

Both points are addressed in https://github.com/servo/servo/issues/31182

Regarding https://github.com/servo/servo/issues/31242, on which 2 above depends: I think merging this initial work is an excellent way to get feedback while trying to fix the event-loop in a follow-up, see the [currently failing test](https://github.com/servo/servo/blob/e15262c5a03f78f480282900d6828be19774ced3/tests/wpt/tests/resize-observer/eventloop.html#L83).

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,262 @@
+}
+
+/// <https://drafts.csswg.org/resize-observer/#calculate-box-size>
+fn calculate_box_size(target: &Element, observed_box: &ResizeObserverBoxOptions) -> Rect<Au> {
```
@mrobinson this is the entry point for the layout calcs.

Loirooriol commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,262 @@
+    ) -> DomRoot<ResizeObserver> {
+        let rooted_observer = ResizeObserver::new(window, proto, callback);
+        let document = window.Document();
+        document.add_resize_observer(&rooted_observer);
```
This is per spec, but has memory problems: https://bugzilla.mozilla.org/show_bug.cgi?id=1596992

I don't remember how Blink managed to avoid that, but I think you should either do the same or disobey the spec like Gecko and WebKit do: https://github.com/w3c/csswg-drafts/issues/4518

gterzian replied:
Thanks for pointing it out, there is a note over at https://github.com/servo/servo/pull/31108/files#diff-b6bdaf97b920500ab94dea9827cda04ff8136b0f23bd0c515d9195c6be31a820R439

and I'll add the info you mention to it...

gterzian commented:
@mrobinson Ping for the review items we discussed...

mrobinson commented:
> @mrobinson Ping for the review items we discussed...

I've just posted #31696 which should allow using the correct bounding rectangle for this API.

mrobinson reviewed (changes requested):
I think this just needs to use the new API for getting the boundaries of the first box for an element.

gterzian commented on components/script/dom/document.rs:
```diff
@@ -2893,6 +2901,61 @@ impl Document {
     }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-resizeobserver>
+    pub fn add_resize_observer(&self, resize_observer: &ResizeObserver) {
```
Why? It seems unnecessary to me since the whole `dom` module is not public.

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,262 @@
+            }
+            // #create-and-populate-a-resizeobserverentry
+            let box_size = calculate_box_size(target, &observation.observed_box);
+            // TODO: writing-mode aware.
```
Yep, cleaned-up these.

jdm commented on components/script/script_thread.rs:
```diff
@@ -1971,6 +1971,28 @@ impl ScriptThread {
+            let _realm = enter_realm(&*document);
+
+            let mut depth = Default::default();
+            loop {
```
Can we use a while here instead?

gterzian replied:
Will do.

jdm commented on tests/wpt/meta/resize-observer/observe.html.ini:
```diff
@@ -0,0 +1,4 @@
+[observe.html]
+  expected: ERROR
```
Do we know what the error is here?

gterzian replied:
Yes, it is a missing API. See https://github.com/servo/servo/issues/31182

jdm commented on tests/wpt/meta/resize-observer/notify.html.ini:
```diff
@@ -0,0 +1,7 @@
+[notify.html]
+  expected: ERROR
```
Do we know what the error is here?

gterzian commented on components/script/dom/document.rs:
```diff
@@ -2920,6 +2931,61 @@ impl Document {
+    /// <https://drafts.csswg.org/resize-observer/#broadcast-active-resize-observations>
+    pub fn broadcast_active_resize_observations(&self) -> ResizeObservationDepth {
+        let mut shallowest = ResizeObservationDepth::max();
+        // Breaking potential re-borrow cycle.
```
When we call the observer callback, new observers can be registered, which would could a double-borrow(will add a more detailed comment).

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,262 @@
+                &[&*observer_size],
+            );
+            entries.push(entry);
+            *observation.last_reported_sizes.borrow_mut() = vec![size_impl];
```
Will do.

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,262 @@
+                *shallowest_target_depth = target_depth;
+            }
+        }
+        let _ = self
```
I've taken another look at all uses of `Call__` in the codebase, and we ignore all return values.

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,262 @@
+}
+
+/// State machine equivalent of active and skipped observations.
+#[derive(Default, JSTraceable, MallocSizeOf)]
```
ok

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,262 @@
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-observedbox>
+    observed_box: ResizeObserverBoxOptions,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-lastreportedsizes>
+    last_reported_sizes: RefCell<Vec<ResizeObserverSizeImpl>>,
```
I guess not, will look into.

gterzian commented:
@jdm @Loirooriol @emilio @mrobinson As mentioned above, I would like to ask for a final review and try to merge this one, so that this would give us the basic infra(now integrated with "update the rendering" in the event-loop), and that remaining layout calculations, as well as a few other points noted in https://github.com/servo/servo/issues/31182, can be left as follow-ups.

jdm reviewed (approved):
Looks great! Thanks!

gterzian commented:
@mrobinson I think the PR needs an approval from you as well. I have addressed your suggestions at the time, but I haven't added the additional layout calculations because I think it's a good follow-up(see comment above).

gterzian commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,277 @@
+                .pop()
+                .unwrap_or_else(|| Rect::zero())
+        },
+        // TODO#31182): add support for border box, and device pixel size, calculations.
```
Thank you.

mrobinson commented on components/script/dom/document.rs:
```diff
@@ -456,6 +457,15 @@ pub struct Document {
+    ///
+    /// Note: we are storing, but never removing, resize observers.
+    /// The lifetime of resize observers is specified at
+    /// https://drafts.csswg.org/resize-observer/#lifetime.
+    /// But implementing it comes with known problems:
+    /// - https://bugzilla.mozilla.org/show_bug.cgi?id=1596992
+    /// - https://github.com/w3c/csswg-drafts/issues/4518.
```
URLs need to be wrapped in rustdoc comments or it will cause a warning:

```suggestion
    /// <https://drafts.csswg.org/resize-observer/#lifetime>.
    /// But implementing it comes with known problems:
    /// - <https://bugzilla.mozilla.org/show_bug.cgi?id=1596992>
    /// - <https://github.com/w3c/csswg-drafts/issues/4518>
```

mrobinson commented on components/script/dom/node.rs:
```diff
@@ -2110,6 +2110,11 @@ impl Node {
         }
         node.owner_doc().remove_script_and_layout_blocker();
+
+        // Note a rendering opportunity
+        let window = window_from_node(parent);
+        let pipeline_id = window.pipeline_id();
+        ScriptThread::note_rendering_opportunity(pipeline_id);
```
```suggestion
        ScriptThread::note_rendering_opportunity(window_from_node(parent).pipeline_id());
```

mrobinson commented on components/script/dom/node.rs:
```diff
@@ -2233,6 +2238,11 @@ impl Node {
             MutationObserver::queue_a_mutation_record(parent, mutation);
         }
         parent.owner_doc().remove_script_and_layout_blocker();
+
+        // Note a rendering opportunity
+        let window = window_from_node(parent);
+        let pipeline_id = window.pipeline_id();
+        ScriptThread::note_rendering_opportunity(pipeline_id);
```
```suggestion
        ScriptThread::note_rendering_opportunity(window_from_node(parent).pipeline_id());
```

mrobinson commented on components/script/script_thread.rs:
```diff
@@ -1967,6 +1984,11 @@ impl ScriptThread {
                 let _realm = enter_realm(&**document);
                 document.maybe_queue_document_completion();
+
+                // Document load is a rendering opportunity.
+                let window = document.window();
+                let pipeline_id = window.pipeline_id();
+                ScriptThread::note_rendering_opportunity(pipeline_id);
```
```suggestion
               // Document load is a rendering opportunity.
                ScriptThread::note_rendering_opportunity(document.window().pipeline_id());
```

mrobinson commented on components/script/dom/document.rs:
```diff
@@ -2920,6 +2930,63 @@ impl Document {
+        self.resize_observers
+            .borrow()
+            .iter()
+            .any(|obs| obs.has_skipped_resize_observations())
```
obs isn't really a well known abbreviation, so I would avoid it:

```suggestion
            .any(|observer| observer.has_skipped_resize_observations())
```

mrobinson commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,277 @@
+        self.observation_targets
+            .borrow()
+            .iter()
+            .any(|(obs, _)| obs.state == ObservationState::Skipped)
```
```suggestion
            .any(|(observeration, _)| observation.state == ObservationState::Skipped)
```

mrobinson commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,277 @@
+            entries.push(entry);
+
+            // Note: this is safe because an observation is
+            // initialized with one reported size(zero).
```
```suggestion
            // initialized with one reported size (zero).
```

mrobinson commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,277 @@
+        }
+        let _ = self
+            .callback
+            .Call__(entries, self, ExceptionHandling::Report);
```
Should this only happen when the size changes? I'm a bit unclear on the mechanism here because is_active() checks to see if the size has changed, but it seems this code both updates the size and then calls the resize observer callbacks. I don't see how the code avoids that if the size changes.

It also seems like `is_active()` is calculating the size but not caching it anywhere...meaning it updates here as well. It's not a cheap calculating right now, unfortunately, so that seems a little wasteful. Perhaps I misunderstand this code though...?

gterzian replied:
> Should this only happen when the size changes?

This only happens for active observations, so that means size has changed plus observation depth is right. The code here does not update the size.

> It's not a cheap calculating right now, unfortunately, so that seems a little wasteful.

I have added a caching mechanism, see the last commit.

mrobinson commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,277 @@
+
+        // Note: noting a rendering opportunity here is necessary
+        // to make /resize-observer/iframe-same-origin.html PASS.
+        let window = window_from_node(target);
+        let pipeline_id = window.pipeline_id();
+        ScriptThread::note_rendering_opportunity(pipeline_id);
```
```suggestion
        ScriptThread::note_rendering_opportunity( window_from_node(target).pipeline_id());
```

mrobinson commented on components/script/dom/resizeobserver.rs:
```diff
@@ -0,0 +1,277 @@
+    fn is_active(&self, target: &Element) -> bool {
+        let last_reported_size = self.last_reported_sizes[0];
+        let box_size = calculate_box_size(target, &self.observed_box);
+        let width = box_size.width().to_f64_px();
+        let height = box_size.height().to_f64_px();
+        (
+            last_reported_size.inline_size(),
+            last_reported_size.block_size(),
+        ) != (width, height)
+    }
```
I think the follow is shorter and clearer:
```suggestion
        box_size.width().to_f64_px() != last_reported_size.inline_size() ||
            box_size.height().to_f64_px() != last_reported_size.block_size()
```

mrobinson reviewed (changes requested):
Just a couple nits, but I don't see any big issues with this change. Hopefully we can land it in the next couple days. Sorry for the big delays!

mrobinson commented:
@gterzian Looks good, though please squash your changes before landing or else the commit message will contain a lot of noise. Thanks you!

gterzian commented:
@Loirooriol @jdm @mrobinson thank you for your reviews.
